docs: document projects and plugins via guides and more method docs (#368)
ApeWorX/ape: docs/userguides/plugins.md

[project template](https://github.com/ApeWorX/project-template) for initializing your plugin. **NOTE**: this template 
is primarily designed for plugins built within the ApeWorX team organization; not everything may apply. It is okay to 
delete anything that does not work or that you don't find helpful. The template may be good to follow if you want to 
keep your plugin of similar quality to plugins developed by the ApeWorX team.

## Implementing API Classes

API classes (classes from the `ape.api` namespace) are primary composed of abstract methods and properties that plugins 
must implement. A benefit of the plugin system is that each plugin can implement these however they need, so long as 
they conform to the API interface. Two plugins with the same API may do entirely different things and yet be 
interchangeable in their usage.

To implement an API, import its class and use it as a base-class in your implementation class. **WARNING**: The plugin 
will fail to work properly if you do not implement all the abstract methods.

```python
from ape.api import ProviderAPI

class MyProvider(ProviderAPI):
  """Implement the abstracts methods here."""

> ```

Reviewer: Not necessary, but maybe show one abstract method in the `ProviderAPI` being overridden?

### Registering API Classes

Once you have finished implementing your API classes, you need to register them using the `@plugins.register`
decorator.

```python
from ape import plugins

# Here, we register our provider plugin so we can use it in 'ape'.
@plugins.register(plugins.ProviderPlugin)
def providers():
    # NOTE: 'MyProvider' defined in a prior code-block.
    yield "ethereum", "development", MyProvider
```

This decorator hooks into ape core and ties everything together by looking for all local installed site-packages that 

> start with `ape_`. Then, it will loop through these potential `ape` plugins and see which ones have created a plugin 

Author: I will add a note earlier on about this

type registration. If the plugin type registration is found, then `ape` knows this package is a plugin and attempts to 
process it according to its registration interface.

### CLI Plugins

The `ape` CLI is built using the python package [click](https://palletsprojects.com/p/click/). To create a CLI plugin,
create any type of `click` command (such as a `click.group` or a `click.command`).

`_cli.py`:

```python
import click

@click.group
def cli():
    """My custom commands."""

@cli.command()
def my_sub_cmd():
    """My subcommand."""
```

Then, register it using `entrypoints`, which is a built-in python registry of items declared in `setup.py`.

`setup.py`:

```python
...
entry_points={
    "ape_cli_subcommands": [
        "ape_myplugin=ape_myplugin._cli:cli",
    ],
},
...
```

**NOTE**: Typically, a `_cli.py` module is used instead of a `__init__.py` module for the location of the Click CLI 
group because it is logically separate from the Python module loading process. If you try to define them together and 
use ape as a library as well, there is a race condition in the loading process that will prevent the cli plugin from 
working.

> For common `ape-click` usages, use the `ape.cli` namespace. For example, use the `@existing_alias_argument()`

Author: Yes! I believe it is and can do that for sure

> For common `ape-click` usages, use the `ape.cli` namespace. For example, use the `@existing_alias_argument()`

Author: I also did the same for `register()`

decorator when you need a CLI argument for specifying an existing account alias:

```python
import click
from ape.cli import existing_alias_argument

@click.command()
@existing_alias_argument()
def my_cmd(alias):
  click.echo(f"{alias} is an existing account!")
```

## Using Plugins

Once you have finished implementing and registering your API classes, they will now be part of `ape`. For example, 
if you implemented the `AccountAPI`, you can now use accounts created from this plugin. The top-level `ape` manager 

> classes are indifferent about the plugin their sources originated from.

Reviewer: ```suggestion
classes are indifferent about the source of the plugin.
```
